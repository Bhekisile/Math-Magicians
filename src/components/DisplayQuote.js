import { useEffect, useState } from 'react';

function DisplayQuote() {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      try {
        const res = await fetch('https://api.api-ninjas.com/v1/quotes?category=faith',
          {
            headers: {
              'X-Api-Key': '5IxClj7l7PEht3AA8nJbiQ==CjzPZ7VSPNDaBiwc',
            },
          });
        if (res.ok) {
          const result = await res.json();
          setData(result[0]);
        } else {
          setError('Failed to fetch quote');
        }
      } catch (error) {
        setError('Failed to fetch quote');
      }
      setIsLoading(false);
    };
    fetchData();
  }, []);

  if (error) {
    return (
      <div className="quote error">
        Error:
        {error}
      </div>
    );
  }

  return (
    <div>
      {isLoading
        ? (<p className="loader">Loading...</p>)
        : (
          <p className="quote">

            {data.quote}
            -
            {data.author}

          </p>
        )}
    </div>
  );
}

export default DisplayQuote;
