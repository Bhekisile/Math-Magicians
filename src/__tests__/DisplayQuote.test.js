import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import DisplayQuote from '../components/DisplayQuote';
import '@testing-library/jest-dom';

describe('display quote', () => {
  it('render quote', () => {
    const { asFragment } = render(<DisplayQuote />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('displays loading state', () => {
    const { getByText } = render(<DisplayQuote />);
    expect(getByText('Loading...')).toBeDefined();
  });

  test('displays error state', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      ok: false,
      json: () => Promise.resolve({}),
    }));

    const { findByText } = render(<DisplayQuote />);
    const errorElement = await findByText(/Error:/i);
    expect(errorElement).toBeDefined();

    global.fetch.mockRestore();
  });

  test('displays fetched quote', async () => {
    const quote = 'Test quote';

    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve([{ quote, author: 'Test author' }]),
    }));

    const { findByText } = await act(async () => render(<DisplayQuote />));

    const quoteElement = await findByText(new RegExp(quote));
    expect(quoteElement).toBeInTheDocument();

    global.fetch.mockRestore();
  });
});
