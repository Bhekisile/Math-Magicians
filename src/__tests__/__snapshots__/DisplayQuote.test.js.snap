// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display quote render quote 1`] = `
<DocumentFragment>
  <div>
    <p
      class="loader"
    >
      Loading...
    </p>
  </div>
</DocumentFragment>
`;
